AppActionVisualizer/app/src/main/java/com/example/appactionvisualizer/utils/StringUtils.java

```diff
@@ -0,0 +1,95 @@
+package com.example.appactionvisualizer.utils;
+
+import android.content.Context;
+
+import com.example.appactionvisualizer.R;
+
+import static com.example.appactionvisualizer.constants.Constant.URL_PARAMETER_INDICATOR;
+
+public class StringUtils {
+
+  /**
+   * @param word1
+   * @param word2
+   * @return Levenshtein distance between two string
+   */
+  public static int distance(String word1, String word2) {
+    int len1 = word1.length();
+    int len2 = word2.length();
+    // if one of the strings is empty
+    if (len1 * len2 == 0)
+      return len1 + len2;
+    int [][] dp = new int[len1 + 1][len2 + 1];
+
+    // init boundaries
+    for (int idx = 0; idx < len1 + 1; idx++) {
+      dp[idx][0] = idx;
+    }
+    for (int idx = 0; idx < len2 + 1; idx++) {
+      dp[0][idx] = idx;
+    }
+
+    // DP compute
+    for (int idx1 = 1; idx1 < len1 + 1; idx1++) {
+      for (int idx2 = 1; idx2 < len2 + 1; idx2++) {
+        int left = dp[idx1 - 1][idx2] + 1;
+        int down = dp[idx1][idx2 - 1] + 1;
+        int left_down = dp[idx1 - 1][idx2 - 1];
+        if (word1.charAt(idx1 - 1) != word2.charAt(idx2 - 1))
+          left_down += 1;
+        dp[idx1][idx2] = Math.min(left, Math.min(down, left_down));
+      }
+    }
+    return dp[len1][len2];
+  }
+
+  /**
+   * @param tobeMatched actions array
+   * @param inputWords input words array
+   * @return a matched "score" by computing matched strings between two arrays
+   */
+  public static int matchScore(String[] tobeMatched, String[] inputWords) {
```

Comment:
seems this is only called within this class and in test, does it have to be public ?

Reply:
I'm not sure to use it as a private method in deep link activity or make it a public. These methods are just string comparison, which have nothing to do with the activity context. And to me a util class contains small and useful methods that might be used someday in the future and that's exactly the use of these methods in StringUtils. So I put these methods in a util class. Correct me if you have better design.